File: cli/src/commands/dispatcher.js
'use strict';

/**
 * onxza dispatcher — Central ONXZA ticket dispatch system.
 *
 * Sub-commands:
 *   status    Show last run, pending deliveries, queue depth, cadence config
 *   run       Trigger dispatcher immediately (one-shot)
 *   dry-run   Scan tickets without sending notifications
 *   logs      Tail dispatcher log
 *
 * The dispatcher runs automatically via OpenClaw cron every 5 minutes.
 * This command provides the health-check surface (ARCHITECTURE.md §6.5).
 *
 * TICKET-20260318-DTP-024 — onxza dispatcher cron automation
 *
 * Imagined by Aaron Gear. Created by Aaron Gear and Marcus Gear (AI Co-Creator).
 * Powered by DevGru US Inc. DBA DevGru Technology Products.
 * Using Powerful Anthropic Models, OpenAI Models, and Local LLMs.
 */

const { Command }    = require('commander');
const { outputJson, isJsonMode } = require('../util');
const fs             = require('fs');
const path           = require('path');
const { execSync, spawnSync } = require('child_process');

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const HOME            = process.env.HOME || os().homedir();
const WORKSPACE       = path.join(HOME, '.openclaw', 'workspace');
const DISPATCHER_DIR  = path.join(WORKSPACE, 'projects', 'onxza', 'dispatcher');
const STATUS_FILE     = path.join(DISPATCHER_DIR, 'status.json');
const LOG_FILE        = path.join(DISPATCHER_DIR, 'dispatcher.log');
const DISPATCHER_PY   = path.join(DISPATCHER_DIR, 'dispatcher.py');

function os() {
  return require('os');
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function readJson(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch {
    return null;
  }
}

function formatTs(isoStr) {
  if (!isoStr) return '—';
  try {
    const d = new Date(isoStr);
    return d.toLocaleString('en-US', { timeZoneName: 'short' });
  } catch {
    return isoStr;
  }
}

function runPython(args) {
  const result = spawnSync('python3', [DISPATCHER_PY, ...args], {
    encoding: 'utf8',
    timeout: 30000,
  });
  return result;
}

// ---------------------------------------------------------------------------
// status sub-command
// ---------------------------------------------------------------------------

function statusAction(options, cmd) {
  const jsonMode = isJsonMode(cmd);

  const status = readJson(STATUS_FILE);

  if (!status) {
    if (jsonMode) {
      outputJson({ status: 'not_run', message: 'Dispatcher has not run yet. Use `onxza dispatcher run` to trigger.' });
    } else {
      console.log('');
      console.log('  ONXZA Dispatcher — Status');
      console.log('  ──────────────────────────────────────────────');
      console.log('  ⚠  No status available. Dispatcher has not run yet.');
      console.log('  Run `onxza dispatcher run` to execute immediately.');
      console.log('');
    }
    return;
  }

  const s = status.last_run_summary || {};

  if (jsonMode) {
    outputJson({
      status:           'ok',
      dispatcher:       status,
    });
    return;
  }

  console.log('');
  console.log('  ONXZA Dispatcher — Health Status');
  console.log('  ──────────────────────────────────────────────────────');
  console.log(`  Version:        ${status.dispatcher_version || '—'}`);
  console.log(`  Last Run:       ${formatTs(status.last_run)}`);
  console.log(`  Total Runs:     ${status.run_count || '—'}`);
  console.log('');
  console.log('  Last Run Summary:');
  console.log(`    Tickets Scanned:   ${s.tickets_scanned ?? '—'}`);
  console.log(`    Tickets Assigned:  ${s.tickets_assigned ?? '—'}`);
  console.log(`    Delivered:         ${s.delivered ?? '—'}`);
  console.log(`    Skipped (cadence): ${s.skipped_cadence ?? '—'}`);
  console.log(`    Errors:            ${s.errors ?? '—'}`);
  console.log(`    Unassigned:        ${s.unassigned ?? '—'}`);
  console.log('');
  console.log(`  Queue Depth:    ${status.queue_depth ?? '—'} pending`);
  console.log(`  Total Sent:     ${status.total_notifications ?? '—'}`);
  console.log('');
  console.log('  Agent Cadences:');
  const cadences = status.agent_cadences || {};
  for (const [role, cadence] of Object.entries(cadences)) {
    console.log(`    ${role.padEnd(20)} ${cadence}`);
  }
  if (status.errors_recent && status.errors_recent.length > 0) {
    console.log('');
    console.log('  Recent Errors:');
    for (const e of status.errors_recent) {
      console.log(`    ⚠  ${e}`);
    }
  }
  console.log('');
}

// ---------------------------------------------------------------------------
// run sub-command
// ---------------------------------------------------------------------------

function runAction(options, cmd) {
  const jsonMode = isJsonMode(cmd);

  if (!fs.existsSync(DISPATCHER_PY)) {
    const msg = `Dispatcher script not found at: ${DISPATCHER_PY}`;
    if (jsonMode) { outputJson({ status: 'error', message: msg }); }
    else { console.log(`\n  ✗ ${msg}\n`); }
    process.exitCode = 1;
    return;
  }

  if (!jsonMode) {
    console.log('');
    console.log('  ONXZA Dispatcher — Running…');
  }

  const result = runPython([]);

  const stdout = (result.stdout || '').trim();
  const stderr = (result.stderr || '').trim();
  const code   = result.status ?? 1;

  if (jsonMode) {
    outputJson({
      status:   code === 0 ? 'ok' : 'error',
      exit_code: code,
      output:   stdout,
      errors:   stderr || null,
    });
    return;
  }

  if (stdout) {
    for (const line of stdout.split('\n')) {
      console.log(`  ${line}`);
    }
  }
  if (stderr) {
    console.log('');
    console.log('  Stderr:');
    for (const line of stderr.split('\n')) {
      console.log(`  ⚠  ${line}`);
    }
  }
  console.log('');
  if (code === 0) {
    console.log('  ✓ Dispatcher run complete.');
  } else {
    console.log(`  ✗ Dispatcher exited with code ${code}`);
    process.exitCode = 1;
  }
  console.log('');
}

// ---------------------------------------------------------------------------
// dry-run sub-command
// ---------------------------------------------------------------------------

function dryRunAction(options, cmd) {
  const jsonMode = isJsonMode(cmd);

  if (!fs.existsSync(DISPATCHER_PY)) {
    const msg = `Dispatcher script not found at: ${DISPATCHER_PY}`;
    if (jsonMode) { outputJson({ status: 'error', message: msg }); }
    else { console.log(`\n  ✗ ${msg}\n`); }
    process.exitCode = 1;
    return;
  }

  const result = runPython(['--dry-run']);
  const stdout = (result.stdout || '').trim();
  const code   = result.status ?? 1;

  if (jsonMode) {
    try {
      // dispatcher --dry-run outputs JSON
      const parsed = JSON.parse(stdout);
      outputJson({ status: 'ok', dry_run: parsed });
    } catch {
      outputJson({ status: code === 0 ? 'ok' : 'error', output: stdout });
    }
    return;
  }

  console.log('');
  console.log('  ONXZA Dispatcher — Dry Run (no notifications sent)');
  console.log('  ──────────────────────────────────────────────────────');

  try {
    const data = JSON.parse(stdout);
    console.log(`  Tickets scanned:  ${data.tickets_scanned}`);
    console.log(`  Tickets assigned: ${data.tickets_assigned}`);
    console.log('');
    if (data.tickets && data.tickets.length > 0) {
      console.log('  Pending dispatch:');
      for (const t of data.tickets) {
        const pri = (t.priority || 'medium').padEnd(8);
        console.log(`    [${pri}] ${t.id}  →  ${t.assigned_to}`);
        if (t.summary) {
          console.log(`             ${t.summary}`);
        }
      }
    } else {
      console.log('  No tickets with assigned agents found.');
    }
  } catch {
    for (const line of stdout.split('\n')) {
      console.log(`  ${line}`);
    }
  }
  console.log('');
}

// ---------------------------------------------------------------------------
// logs sub-command
// ---------------------------------------------------------------------------

function logsAction(options, cmd) {
  const jsonMode = isJsonMode(cmd);
  const lines = options.lines || 50;

  if (!fs.existsSync(LOG_FILE)) {
    if (jsonMode) {
      outputJson({ status: 'not_found', message: 'No dispatcher log yet.' });
    } else {
      console.log('\n  No dispatcher log yet. Run `onxza dispatcher run` first.\n');
    }
    return;
  }

  try {
    const content = fs.readFileSync(LOG_FILE, 'utf8');
    const allLines = content.split('\n').filter(Boolean);
    const tail = allLines.slice(-lines);

    if (jsonMode) {
      outputJson({ status: 'ok', lines: tail });
      return;
    }

    console.log('');
    console.log(`  ONXZA Dispatcher — Last ${lines} log lines`);
    console.log('  ──────────────────────────────────────────────');
    for (const l of tail) {
      console.log(`  ${l}`);
    }
    console.log('');
  } catch (e) {
    if (jsonMode) {
      outputJson({ status: 'error', message: e.message });
    } else {
      console.log(`\n  ✗ Cannot read log: ${e.message}\n`);
    }
    process.exitCode = 1;
  }
}

// ---------------------------------------------------------------------------
// Command definitions
// ---------------------------------------------------------------------------

const dispatcherCmd = new Command('dispatcher')
  .description('ONXZA central dispatcher — ticket scanning and agent notification');

dispatcherCmd
  .command('status')
  .description('Show dispatcher health: last run, queue depth, cadence config')
  .action(statusAction);

dispatcherCmd
  .command('run')
  .description('Trigger dispatcher immediately (scan tickets and send notifications)')
  .action(runAction);

dispatcherCmd
  .command('dry-run')
  .description('Scan open tickets and preview dispatch targets without sending')
  .action(dryRunAction);

dispatcherCmd
  .command('logs')
  .description('Show recent dispatcher log output')
  .option('-n, --lines <count>', 'Number of lines to show', '50')
  .action(logsAction);

module.exports = dispatcherCmd;

File: dispatcher/dispatcher.py
#!/usr/bin/env python3
"""
ONXZA Dispatcher — Central ticket scan and agent notification engine.
Version: 1.1.0

Architecture §6.5: Runs every 5 minutes via OpenClaw cron.
Scans tickets/open/, resolves assigned_to, sends TICKET_ASSIGNED
notifications to each agent session. Tracks delivery state in
dispatcher/state.json to avoid duplicate notifications.

Fixes in v1.1.0:
  - Alias resolution: PascalCase, legacy IDs, and group addresses
  - Exit 0 on partial success (only exit 1 if ALL deliveries fail)
  - Workspace discovery: scan actual ~/.openclaw dirs for matching agent
  - Skip group addresses (all-mga-agents etc) with a WARN, not ERROR
  - Suppress repeated same-ticket errors in log (dedup per run)

Agent polling cadence (ARCHITECTURE.md §6.5):
  - marcus / mg-parent-marcus     : continuous (no cadence gate)
  - CEO Orchestrators             : session-start + every 15 min
  - Department PMs                : session-start + every 30 min
  - Specialists                   : session-start + on-task-completion
  - QualityDirector               : session-start + every 60 min

Imagined by Aaron Gear. Created by Aaron Gear and Marcus Gear (AI Co-Creator).
Powered by DevGru US Inc. DBA DevGru Technology Products.
Using Powerful Anthropic Models, OpenAI Models, and Local LLMs.
"""

import os
import sys
import json
import re
import time
import datetime
import subprocess
from pathlib import Path

# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------

OPENCLAW_DIR     = Path.home() / ".openclaw"
WORKSPACE        = OPENCLAW_DIR / "workspace"
TICKETS_OPEN     = WORKSPACE / "tickets" / "open"
DISPATCHER_DIR   = WORKSPACE / "projects" / "onxza" / "dispatcher"
STATE_FILE       = DISPATCHER_DIR / "state.json"
LOG_FILE         = DISPATCHER_DIR / "dispatcher.log"

# ---------------------------------------------------------------------------
# Agent ID alias table
# ID normalization: map ticket assigned_to values → actual workspace directory suffix
# ---------------------------------------------------------------------------

AGENT_ALIASES = {
    # PascalCase → kebab-case
    "MG_Parent_Marcus":        "mg-parent-marcus",
    "MG_Parent_Orchestrator":  "mg-parent-orchestrator",
    "MG_Parent_AgentDeveloper": "mg-parent-agentdeveloper",
    "MG_Parent_Security":      "mg-parent-security",
    "WDC_Content_PM":          "wdc-content-pm",
    "WDC_Website_PM":          "wdc-website-pm",
    "WDC_Affiliate_PM":        "wdc-affiliate-pm",
    "WDC_CEO":                 "wdc-ceo",
    "WDC_COO":                 "wdc-coo",
    "DTP_ONXZA_PM":            "dtp-onxza-pm",
    "DTP_ONXZA_Router":        "dtp-onxza-router",
    "DTP_ONXZA_Architect":     "dtp-onxza-architect",
    "DTP_CEO":                 "dtp-ceo",
    "DTP_COO":                 "dtp-coo",
    # Legacy/short names
    "marcus":                  "mg-parent-marcus",
    "main":                    "mg-parent-marcus",
    "orchestrator":            "mg-parent-orchestrator",
    # Workspace variant names
    "mg-parent-marcus":        "mg-parent-marcus",
    "workspace":               "mg-parent-marcus",
}

# Group addresses — not deliverable to a single workspace; skip with info log
GROUP_ADDRESSES = re.compile(
    r"^(all-|everyone|broadcast|aaron.gear|Aaron.Gear)", re.I
)

# Human-only recipients — skip gracefully
HUMAN_RECIPIENTS = re.compile(
    r"^(aaron|Aaron Gear|aaron gear|aaron\.gear)$", re.I
)

# ---------------------------------------------------------------------------
# Cadence tier definitions
# ---------------------------------------------------------------------------
# interval_minutes = None  → continuous (always eligible)
# interval_minutes = 0     → on-task-completion (dispatcher always delivers;
#                             agent self-gates via TASK_STATE)

CADENCE_RULES = [
    (re.compile(r"^(main|mg-parent-marcus|marcus|mg-parent-orchestrator|orchestrator)$", re.I),
     "continuous", None),
    (re.compile(r"-(ceo|coo)$", re.I),                    "ceo-coo",         15),
    (re.compile(r"-pm$", re.I),                            "pm",              30),
    (re.compile(r"(-qualitydirector|-quality-director)$", re.I), "quality",  60),
    (re.compile(r"-(verification|qa)$", re.I),             "quality",         60),
    (re.compile(r".*", re.I),                              "specialist",       0),
]

# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------

def log(msg: str, level: str = "INFO") -> None:
    ts = datetime.datetime.now().isoformat(timespec="seconds")
    line = f"[{ts}] [{level}] {msg}"
    print(line, flush=True)
    try:
        DISPATCHER_DIR.mkdir(parents=True, exist_ok=True)
        with open(LOG_FILE, "a") as f:
            f.write(line + "\n")
    except Exception:
        pass


# ---------------------------------------------------------------------------
# State management
# ---------------------------------------------------------------------------

def load_state() -> dict:
    if STATE_FILE.exists():
        try:
            with open(STATE_FILE) as f:
                return json.load(f)
        except Exception as e:
            log(f"State load error — starting fresh: {e}", "WARN")
    return {
        "last_run":            None,
        "run_count":           0,
        "notifications_sent":  {},
        "agent_last_notified": {},
        "errors":              [],
    }


def save_state(state: dict) -> None:
    DISPATCHER_DIR.mkdir(parents=True, exist_ok=True)
    tmp = STATE_FILE.with_suffix(".tmp")
    try:
        with open(tmp, "w") as f:
            json.dump(state, f, indent=2)
        tmp.replace(STATE_FILE)
    except Exception as e:
        log(f"State save error: {e}", "ERROR")


# ---------------------------------------------------------------------------
# Workspace discovery
# ---------------------------------------------------------------------------

def build_workspace_index() -> dict[str, Path]:
    """
    Scan ~/.openclaw/ for workspace-* directories.
    Returns {agent-id: Path} index.
    The main Marcus workspace lives at ~/.openclaw/workspace (no suffix).
    """
    index = {}
    for d in OPENCLAW_DIR.iterdir():
        if not d.is_dir():
            continue
        if d.name == "workspace":
            # Main workspace — maps to mg-parent-marcus and aliases
            index["workspace"] = d
            index["mg-parent-marcus"] = d
            index["marcus"] = d
            index["main"] = d
        elif d.name.startswith("workspace-"):
            agent_id = d.name[len("workspace-"):]
            index[agent_id] = d
    return index


# ---------------------------------------------------------------------------
# Agent ID resolution
# ---------------------------------------------------------------------------

def resolve_agent_id(raw_id: str, workspace_index: dict) -> str | None:
    """
    Resolve a raw assigned_to value to an actual workspace key.
    Returns the resolved agent ID string, or None if unresolvable.
    """
    if not raw_id or not raw_id.strip():
        return None

    raw = raw_id.strip()

    # Skip group addresses
    if GROUP_ADDRESSES.match(raw):
        log(f"  ℹ  Group address skipped: {raw}")
        return None

    # Skip human recipients
    if HUMAN_RECIPIENTS.match(raw):
        log(f"  ℹ  Human recipient skipped (requires_aaron): {raw}")
        return None

    # 1. Direct alias lookup
    resolved = AGENT_ALIASES.get(raw)
    if resolved and resolved in workspace_index:
        return resolved

    # 2. Direct match in workspace index (already kebab-case)
    if raw in workspace_index:
        return raw

    # 3. Normalize PascalCase → kebab-case and retry
    # DTP_ONXZA_Architect → dtp-onxza-architect
    normalized = re.sub(r"[_\s]+", "-", raw).lower()
    if normalized in workspace_index:
        return normalized

    # 4. Underscore → hyphen variant
    hyphenated = raw.replace("_", "-").lower()
    if hyphenated in workspace_index:
        return hyphenated

    return None


# ---------------------------------------------------------------------------
# Ticket parsing
# ---------------------------------------------------------------------------

YAML_FRONTMATTER = re.compile(r"^---\s*\n(.*?)\n---\s*\n", re.DOTALL)


def parse_frontmatter(content: str) -> dict:
    meta = {}
    m = YAML_FRONTMATTER.match(content)
    if not m:
        return meta
    for line in m.group(1).splitlines():
        if ":" not in line:
            continue
        key, _, value = line.partition(":")
        meta[key.strip()] = value.strip()
    return meta


def extract_ticket_id(filename: str) -> str:
    name = Path(filename).stem
    m = re.match(r"(TICKET-[A-Z0-9\-]+|HC2-[A-Z0-9\-]+|[A-Z][A-Z0-9\-]+-[0-9]+)", name, re.I)
    return m.group(1).upper() if m else name.upper()


def load_open_tickets() -> list[dict]:
    tickets = []
    if not TICKETS_OPEN.exists():
        log(f"tickets/open/ not found at {TICKETS_OPEN}", "WARN")
        return tickets

    for path in sorted(TICKETS_OPEN.glob("*.md")):
        try:
            content = path.read_text(encoding="utf-8", errors="replace")
        except Exception as e:
            log(f"Cannot read {path.name}: {e}", "WARN")
            continue

        meta = parse_frontmatter(content)
        ticket_id   = meta.get("id", "") or meta.get("ticket_id", "") or extract_ticket_id(path.name)
        assigned_to = (meta.get("assigned_to", "") or "").strip()
        priority    = (meta.get("priority", "medium") or "medium").strip()

        summary_m = re.search(r"^## Summary\s*\n+(.*?)(?:\n## |\Z)", content, re.DOTALL | re.MULTILINE)
        summary = summary_m.group(1).strip().replace("\n", " ")[:200] if summary_m else ""

        tickets.append({
            "id":          ticket_id,
            "assigned_to": assigned_to,
            "priority":    priority,
            "summary":     summary,
            "filename":    path.name,
            "path":        str(path),
        })

    return tickets


# ---------------------------------------------------------------------------
# Cadence gating
# ---------------------------------------------------------------------------

def cadence_interval(agent_id: str) -> int | None:
    for pattern, _label, interval in CADENCE_RULES:
        if pattern.search(agent_id):
            return interval
    return 0


def agent_is_eligible(agent_id: str, agent_last_notified: dict, now_ts: float) -> bool:
    interval = cadence_interval(agent_id)
    if interval is None:
        return True
    if interval == 0:
        return True
    last_str = agent_last_notified.get(agent_id)
    if not last_str:
        return True
    try:
        last_ts = datetime.datetime.fromisoformat(last_str).timestamp()
    except ValueError:
        return True
    return (now_ts - last_ts) / 60.0 >= interval


# ---------------------------------------------------------------------------
# Notification delivery
# ---------------------------------------------------------------------------

def deliver_notification(agent_id: str, ticket: dict, workspace_index: dict) -> bool:
    workspace = workspace_index.get(agent_id)
    if not workspace:
        log(f"  ✗ No workspace for resolved agent_id={agent_id}", "ERROR")
        return False

    inbox_dir = workspace / "inbox"
    inbox_dir.mkdir(exist_ok=True)

    # DEDUPLICATION: Check if a TICKET_ASSIGNED file for this ticket already exists
    # in the inbox. If so, skip — don't flood the inbox with duplicates.
    ticket_id = ticket['id']
    existing = list(inbox_dir.glob(f"TICKET_ASSIGNED-{ticket_id}-*.md"))
    if existing:
        log(f"  ⏭ Skipped (already in inbox): {agent_id} — {ticket_id}")
        return True  # Not an error — just already delivered

    ts_str = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
    inbox_file = inbox_dir / f"TICKET_ASSIGNED-{ticket_id}-{ts_str}.md"

    payload = build_payload(ticket)
    try:
        inbox_file.write_text(payload, encoding="utf-8")
        log(f"  ✓ Delivered to {agent_id}: {inbox_file.name}")
        return True
    except Exception as e:
        log(f"  ✗ Write failed for {agent_id}: {e}", "ERROR")
        return False


def build_payload(ticket: dict) -> str:
    now = datetime.datetime.now().isoformat(timespec="seconds")
    return f"""# TICKET_ASSIGNED

**Ticket ID:** {ticket['id']}
**Priority:** {ticket['priority']}
**Assigned To:** {ticket['assigned_to']}
**File:** {ticket['filename']}

## Summary
{ticket['summary'] or '(no summary)'}

## Action Required
You have an open ticket assigned to you. Check TASK_STATE in your MEMORY.md.
- If TASK_STATE: IDLE → begin work immediately
- If TASK_STATE: ACTIVE → queue for your next available slot

---
*Dispatched by ONXZA Dispatcher v1.1.0 at {now}*
*Imagined by Aaron Gear. Created by Aaron Gear and Marcus Gear (AI Co-Creator). Powered by DevGru Technology Products.*
"""


# ---------------------------------------------------------------------------
# Health status
# ---------------------------------------------------------------------------

def write_health_status(state: dict, run_summary: dict) -> None:
    DISPATCHER_DIR.mkdir(parents=True, exist_ok=True)
    status = {
        "dispatcher_version":  "1.1.0",
        "last_run":            state.get("last_run"),
        "run_count":           state.get("run_count", 0),
        "last_run_summary":    run_summary,
        "queue_depth":         0,
        "total_notifications": len(state.get("notifications_sent", {})),
        "agent_cadences": {
            "marcus/orchestrator":  "continuous",
            "ceo-orchestrator":     "every 15 min",
            "pm":                   "every 30 min",
            "quality-director":     "every 60 min",
            "specialist":           "on-task-completion",
        },
        "errors_recent": state.get("errors", [])[-10:],
    }
    tmp = (DISPATCHER_DIR / "status.json").with_suffix(".tmp")
    with open(tmp, "w") as f:
        json.dump(status, f, indent=2)
    tmp.replace(DISPATCHER_DIR / "status.json")


# ---------------------------------------------------------------------------
# Main dispatch loop
# ---------------------------------------------------------------------------

def run_dispatch(dry_run: bool = False) -> dict:
    log("─" * 60)
    log(f"ONXZA Dispatcher v1.1.0 {'[DRY RUN] ' if dry_run else ''}starting")

    state   = load_state()
    now_ts  = time.time()
    now_iso = datetime.datetime.now().isoformat(timespec="seconds")

    state["run_count"] = state.get("run_count", 0) + 1
    state["last_run"]  = now_iso

    # Build workspace index once
    workspace_index = build_workspace_index()
    log(f"Workspace index: {len(workspace_index)} entries")

    tickets = load_open_tickets()
    log(f"Found {len(tickets)} open ticket(s)")

    assigned      = [t for t in tickets if t["assigned_to"]]
    unassigned    = len(tickets) - len(assigned)
    if unassigned:
        log(f"{unassigned} ticket(s) with no assigned_to — skipping")

    delivered    = 0
    skipped      = 0
    errors       = 0
    unresolvable = 0

    notifications_sent  = state.setdefault("notifications_sent", {})
    agent_last_notified = state.setdefault("agent_last_notified", {})

    # Dedup errors within this run (avoid log spam for same broken agent)
    seen_unresolvable = set()

    for ticket in assigned:
        tid      = ticket["id"]
        raw_id   = ticket["assigned_to"]

        resolved = resolve_agent_id(raw_id, workspace_index)
        if resolved is None:
            if raw_id not in seen_unresolvable:
                log(f"  ⚠  Cannot resolve agent_id '{raw_id}' — no workspace found", "WARN")
                seen_unresolvable.add(raw_id)
            unresolvable += 1
            continue

        # Cadence gate
        if not agent_is_eligible(resolved, agent_last_notified, now_ts):
            interval = cadence_interval(resolved)
            log(f"  ⏳ Cadence skip: {resolved} (next in {interval}min) — {tid}")
            skipped += 1
            continue

        log(f"  → Dispatching {tid} to {resolved} (raw: {raw_id}) [{ticket['priority']}]")

        if dry_run:
            delivered += 1
            continue

        ok = deliver_notification(resolved, ticket, workspace_index)
        if ok:
            delivered += 1
            agent_last_notified[resolved] = now_iso
            notifications_sent[tid] = {
                "sent_at":         now_iso,
                "agent_id":        resolved,
                "raw_agent_id":    raw_id,
                "priority":        ticket["priority"],
                "delivery_status": "delivered",
            }
        else:
            errors += 1
            err_msg = f"Delivery failed: {tid} → {resolved}"
            errs = state.setdefault("errors", [])
            errs.append(f"{now_iso}: {err_msg}")
            state["errors"] = errs[-50:]

    run_summary = {
        "run_at":            now_iso,
        "tickets_scanned":   len(tickets),
        "tickets_assigned":  len(assigned),
        "delivered":         delivered,
        "skipped_cadence":   skipped,
        "unresolvable":      unresolvable,
        "errors":            errors,
        "unassigned":        unassigned,
        "dry_run":           dry_run,
    }

    log(f"Run complete — delivered:{delivered} cadence_skip:{skipped} "
        f"unresolvable:{unresolvable} write_errors:{errors}")

    if not dry_run:
        save_state(state)
        write_health_status(state, run_summary)

    return run_summary


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(
        description="ONXZA Dispatcher v1.1.0 — scan open tickets and notify assigned agents"
    )
    parser.add_argument("--dry-run", action="store_true",
                        help="Scan and report without sending notifications")
    parser.add_argument("--status", action="store_true",
                        help="Print health status JSON and exit")
    args = parser.parse_args()

    if args.status:
        status_path = DISPATCHER_DIR / "status.json"
        if status_path.exists():
            print(status_path.read_text())
        else:
            print(json.dumps({"error": "No dispatcher status yet. Run dispatcher once first."}, indent=2))
        sys.exit(0)

    summary = run_dispatch(dry_run=args.dry_run)

    if args.dry_run:
        print(json.dumps(summary, indent=2))

    # Exit 0 unless all deliveries failed (hard error). Partial success = ok.
    total_attempted = summary["delivered"] + summary["errors"]
    if total_attempted > 0 and summary["errors"] == total_attempted:
        sys.exit(1)
    sys.exit(0)
